Sources/MapboxDirections/RouteResponse.swift

```diff
@@ -58,12 +58,28 @@ extension RouteResponse: Codable {
             routes = try decoder.decode([Route].self, from: matchesData)
         }
         
-        var waypoints: [Waypoint]?
+        var waypoints: [Route.Waypoint]?
         
         if let tracepoints = response.tracepoints {
             let filtered = tracepoints.compactMap { $0 }
             let tracepointsData = try encoder.encode(filtered)
-            waypoints = try decoder.decode([Waypoint].self, from: tracepointsData)
+            waypoints = try decoder.decode([Route.Waypoint].self, from: tracepointsData)
+            
+            // Push tracepoints down to individual legs to reflect waypoint_index.
+            if let routes = routes {
+                let sorted = zip(filtered, waypoints!).sorted {
+                    $0.0.waypointIndex < $1.0.waypointIndex
+                }
+                for (index, route) in routes.enumerated() {
+                    let legSeparators = sorted.filter {
+                        return $0.0.matchingIndex == index
+                    }.map { $0.1 }
+                    
+                    assert(legSeparators.count == route.legs.count + 1)
+                    
+                    route.legSeparators = legSeparators
+                }
+            }
```

Comment:
This part is how I understand [Map Matching Doc](https://docs.mapbox.com/api/navigation/#map-matching). The idea is to set leg separators explicitly via `matchings_index` and `waypoint_index`. Not sure if it is required though or will have any effect at all since `routes` and `waypoints` are passed to the user as public properties and waypoints are not modified in this code.
A help/review would be required here.